### ISSUE TITLE
-DT -e 'use warnings;' crashes

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#8117]
Migrated from [rt.perl.org#37190](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=37190) (status was 'resolved')

Searchable as RT37190$

**Comment by p5pRT** [comment]
### From @nwc10

#### Created by @nwc10



Using \-DT with use warnings; goes bang on OS X\. I was able to get "out of
memory" failures on FreeBSD\, but everything works on x86/Linux\, at least for
me\. I guess x86/Linux is just lucky \- this seems to be a real bug\, although
quite where\, I'm not sure\.

Starting gdb as gdb \-\-args \./perl \-Ilib \-DT \-e 'use warnings;'
and running the program gives&#8203;:

Program received signal EXC\_BAD\_ACCESS\, Could not access memory\.
0xffff8ab0 in \_\_memcpy \(\)
\(gdb\) where
\#0  0xffff8ab0 in \_\_memcpy \(\)
\#1  0x00065390 in Perl\_sv\_catpvn\_flags \(my\_perl=0x800200\, dsv=0x815060\, sstr=0x409870 "\\\\x55\\\\x55\\\\x55\\\\x55\\\\x55\\\\x55\\\\x55\\\\x55\\\\x55\\\\x55\\\\x55\\\\x15"\, slen=4235230\, flags=2\) at sv\.c&#8203;:4723
\#2  0x0013eadc in S\_tokereport \(my\_perl=0x800200\, s=0x81384e "\, \# \[0\.\.46\]\\n"\, rv=40\) at toke\.c&#8203;:327
\#3  0x00146628 in Perl\_yylex \(my\_perl=0x800200\) at toke\.c&#8203;:2517
\#4  0x0019304c in Perl\_yyparse \(my\_perl=0x800200\) at perly\.c&#8203;:412
\#5  0x000efa90 in S\_doeval \(my\_perl=0x800200\, gimme=0\, startop=0x0\, outside=0x0\, seq=0\) at pp\_ctl\.c&#8203;:2954
\#6  0x000f20ac in Perl\_pp\_require \(my\_perl=0x800200\) at pp\_ctl\.c&#8203;:3411
\#7  0x00173420 in Perl\_runops\_debug \(my\_perl=0x800200\) at dump\.c&#8203;:1597
\#8  0x00032500 in S\_call\_body \(my\_perl=0x800200\, myop=0xbfffe540\, is\_eval=0 '\\0'\) at perl\.c&#8203;:2617
\#9  0x00031f70 in Perl\_call\_sv \(my\_perl=0x800200\, sv=0x810220\, flags=6\) at perl\.c&#8203;:2546
\#10 0x00038c2c in S\_call\_list\_body \(my\_perl=0x800200\, cv=0x810220\) at perl\.c&#8203;:5123
\#11 0x0003861c in Perl\_call\_list \(my\_perl=0x800200\, oldscope=2\, paramList=0x8102f0\) at perl\.c&#8203;:5062
\#12 0x000212c4 in Perl\_newATTRSUB \(my\_perl=0x800200\, floor=112\, o=0x405d70\, proto=0x0\, attrs=0x0\, block=0x405e00\) at op\.c&#8203;:4541
\#13 0x0001b33c in Perl\_utilize \(my\_perl=0x800200\, aver=1\, floor=112\, version=0x0\, idop=0x405b60\, arg=0x0\) at op\.c&#8203;:3079
\#14 0x001945ac in Perl\_yyparse \(my\_perl=0x800200\) at perly\.y&#8203;:382
\#15 0x00030730 in S\_parse\_body \(my\_perl=0x800200\, env=0x0\, xsinit=0x2c04 \<xs\_init>\) at perl\.c&#8203;:2133
\#16 0x0002edc0 in perl\_parse \(my\_perl=0x800200\, xsinit=0x2c04 \<xs\_init>\, argc=5\, argv=0xbffff6fc\, env=0x0\) at perl\.c&#8203;:1540
\#17 0x00002b74 in main \(argc=5\, argv=0xbffff6fc\, env=0xbffff714\) at perlmain\.c&#8203;:101
\(gdb\) up
\#1  0x00065390 in Perl\_sv\_catpvn\_flags \(my\_perl=0x800200\, dsv=0x815060\, sstr=0x409870 "\\\\x55\\\\x55\\\\x55\\\\x55\\\\x55\\\\x55\\\\x55\\\\x55\\\\x55\\\\x55\\\\x55\\\\x15"\, slen=4235230\, flags=2\) at sv\.c&#8203;:4723
4723        Move\(sstr\, SvPVX\(dsv\) \+ dlen\, slen\, char\);
\(gdb\) p sstr
$1 = 0x409870 "\\\\x55\\\\x55\\\\x55\\\\x55\\\\x55\\\\x55\\\\x55\\\\x55\\\\x55\\\\x55\\\\x55\\\\x15"
\(gdb\) p slen
$2 = 4235230

I suspect that this "length" has something to do with the "out of memory" on
FreeBSD\. I've no idea what the pattern in the memory pointed to by sstr
represents\.

Nicholas Clark

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=low

Site configuration information for perl vv5.9.3:

Configured by nick at Sat Sep 17 18:27:53 BST 2005.

Summary of my perl5 (revision 5 version 9 subversion 3) configuration:
  Platform:
    osname=darwin, osvers=7.9.0, archname=darwin-thread-multi-2level
    uname='darwin ship-in-a-bottle 7.9.0 darwin kernel version 7.9.0: wed mar 30 20:11:17 pst 2005; root:xnuxnu-517.12.7.obj~1release_ppc power macintosh powerpc '
    config_args='-Dusedevel=y -Dcc=ccache gcc -Dld=gcc -Ubincompat5005 -Uinstallusrbinperl -Dcf_email=nick@ccl4.org -Dperladmin=nick@ccl4.org -Doptimize=-g -Dusethreads=y -Uuse64bitint -Duselargefiles -Dprefix=~/Sandpit/blead25436 -de'
    hint=recommended, useposix=true, d_sigaction=define
    useithreads=define, usemultiplicity=define
    useperlio=define, d_sfio=undef, uselargefiles=define, usesocks=undef
    use64bitint=undef, use64bitall=undef, uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='ccache gcc', ccflags ='-fno-common -DPERL_DARWIN -no-cpp-precomp -DDEBUGGING -fno-strict-aliasing -pipe',
    optimize='-g',
    cppflags='-no-cpp-precomp -fno-common -DPERL_DARWIN -no-cpp-precomp -DDEBUGGING -fno-strict-aliasing -pipe'
    ccversion='', gccversion='3.3 20030304 (Apple Computer, Inc. build 1666)', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=4321
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=8
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=8, prototype=define
  Linker and Libraries:
    ld='env MACOSX_DEPLOYMENT_TARGET=10.3 cc', ldflags =' -L/usr/local/lib'
    libpth=/usr/local/lib /usr/lib
    libs=-ldbm -ldl -lm -lc
    perllibs=-ldl -lm -lc
    libc=/usr/lib/libc.dylib, so=dylib, useshrplib=false, libperl=libperl.a
    gnulibc_version=''
  Dynamic Linking:
    dlsrc=dl_dyld.xs, dlext=bundle, d_dlsymun=undef, ccdlflags=' '
    cccdlflags=' ', lddlflags=' -bundle -undefined dynamic_lookup -L/usr/local/lib'

Locally applied patches:
    


@INC for perl vv5.9.3:
    lib
    /sw/lib/perl5
    /sw/lib/perl5/darwin
    /Users/nick/Sandpit/blead25436/lib/perl5/5.9.3/darwin-thread-multi-2level
    /Users/nick/Sandpit/blead25436/lib/perl5/5.9.3
    /Users/nick/Sandpit/blead25436/lib/perl5/site_perl/5.9.3/darwin-thread-multi-2level
    /Users/nick/Sandpit/blead25436/lib/perl5/site_perl/5.9.3
    /Users/nick/Sandpit/blead25436/lib/perl5/site_perl
    .


Environment for perl vv5.9.3:
    DYLD_LIBRARY_PATH (unset)
    HOME=/Users/nick
    LANG (unset)
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/Users/nick/bin:/sw/bin:/sw/sbin:/bin:/sbin:/usr/bin:/usr/sbin:/usr/X11R6/bin:/usr/local/sbin:/sbin:/usr/sbin
    PERL5LIB=/sw/lib/perl5:/sw/lib/perl5/darwin
    PERL_BADLANG (unset)
    SHELL=/bin/bash

```
</details>

**Comment by p5pRT** [comment]
### From rick@bort.ca

On Sat\, Sep 17\, 2005 at 11&#8203;:31&#8203;:11AM \-0700\, Nicholas Clark wrote&#8203;:
> \(gdb\) up
> \#1  0x00065390 in Perl\_sv\_catpvn\_flags \(my\_perl=0x800200\, dsv=0x815060\, sstr=0x409870 "\\\\x55\\\\x55\\\\x55\\\\x55\\\\x55\\\\x55\\\\x55\\\\x55\\\\x55\\\\x55\\\\x55\\\\x15"\, slen=4235230\, flags=2\) at sv\.c&#8203;:4723
> 4723        Move\(sstr\, SvPVX\(dsv\) \+ dlen\, slen\, char\);
> \(gdb\) p sstr
> $1 = 0x409870 "\\\\x55\\\\x55\\\\x55\\\\x55\\\\x55\\\\x55\\\\x55\\\\x55\\\\x55\\\\x55\\\\x55\\\\x15"
> \(gdb\) p slen
> $2 = 4235230
> 
> I suspect that this "length" has something to do with the "out of memory" on
> FreeBSD\. I've no idea what the pattern in the memory pointed to by sstr
> represents\.

The pattern is the bits for which warning categories are set\, or
WARN\_ALLstring\.  The length should be WARNsize \(12\)\.

\-\- 
Rick Delaney
rick@&#8203;bort\.ca

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @rgs

Nicholas Clark \(via RT\) wrote&#8203;:
> Using \-DT with use warnings; goes bang on OS X\. I was able to get "out of
> memory" failures on FreeBSD\, but everything works on x86/Linux\, at least for
> me\. I guess x86/Linux is just lucky \- this seems to be a real bug\, although
> quite where\, I'm not sure\.
> 
> Starting gdb as gdb \-\-args \./perl \-Ilib \-DT \-e 'use warnings;'
> and running the program gives&#8203;:

Moreover\, this makes valgrind crash\, giving this warning before hanging
and eating all CPU :

==13290== Warning&#8203;: set address range perms&#8203;: large range 171544052\, a 0\, v 1

**Comment by p5pRT** [comment]
### From @iabyn

On Sat\, Sep 17\, 2005 at 11&#8203;:31&#8203;:11AM \-0700\, Nicholas Clark wrote&#8203;:
> Using \-DT with use warnings; goes bang on OS X\. I was able to get "out of
> memory" failures on FreeBSD\, but everything works on x86/Linux\, at least for
> me\. I guess x86/Linux is just lucky \- this seems to be a real bug\, although
> quite where\, I'm not sure\.

It looks like its crashing in code I wrote\, so its probably my fault\.
If you don't mind waiting for a couple of weeks\, I'll try to have a look
somretime\.

\-\- 
Thank God I'm an atheist\.\.\.\.\.

**Comment by p5pRT** [comment]
### From @iabyn

On Sat\, Sep 17\, 2005 at 11&#8203;:31&#8203;:11AM \-0700\, Nicholas Clark wrote&#8203;:
> Using \-DT with use warnings; goes bang on OS X\. I was able to get "out of
> memory" failures on FreeBSD\, but everything works on x86/Linux\, at least for
> me\. I guess x86/Linux is just lucky \- this seems to be a real bug\, although
> quite where\, I'm not sure\.

Fixed by change \#25674\. My debugging code had made some unwarranted
assumptions about the values of various buffer ptrs at various times\.

In fact I've bypassed the problem by completely rejigging the debugging
output\. \-DT now shows the line number\, state and curent buffer at entry to
yylex\(\) rather than showing the buffer at exit\. It now also escapes chars
when printing\, eg newline is displayed as \\n\.

It also displays the value of a returned OP\_CONST\.

The format is a bit more concise\, and it prints a blank line after each
return value\.

Here's a before and after sample of \-DT&#8203;:



---
Before&#8203;:

\./perl \-DT \-e'$a=$b\+f\(3\)'
\#\#\# LEX\_NORMAL
\#\#\# Tokener expecting STATE at \[
;\]
\#\#\# \<== '$' at line 1 \[$a\]
\#\#\# LEX\_NORMAL
\#\#\# Tokener saw identifier '$a'
\#\#\# \<== WORD\(opval=op\_const\) at line 1 \[$a\]
\#\#\# LEX\_NORMAL
\#\#\# Tokener expecting OPERATOR at \[=$b\+f\(3\)
\]
\#\#\# \<== ASSIGNOP\(ival=op\_null\) at line 1 \[$a\]
\#\#\# LEX\_NORMAL
\#\#\# Tokener expecting TERM at \[$b\+f\(3\)
\]
\#\#\# \<== '$' at line 1 \[$b\]
\#\#\# LEX\_NORMAL
\#\#\# Tokener saw identifier '$b'
\#\#\# \<== WORD\(opval=op\_const\) at line 1 \[$b\]
\#\#\# LEX\_NORMAL
\#\#\# Tokener expecting OPERATOR at \[\+f\(3\)
\]
\#\#\# \<== ADDOP\(ival=op\_add\) at line 1 \[$b\]
\#\#\# LEX\_NORMAL
\#\#\# Tokener expecting TERM at \[f\(3\)
\]
\#\#\# \<== '&' at line 1 \[f\]
\#\#\# LEX\_KNOWNEXT
\#\#\# Next token after '\(3\)
' was known\, type 258
\#\#\# \<== WORD\(opval=op\_const\) at line 1 \[f\]
\#\#\# LEX\_NORMAL
\#\#\# Tokener expecting OPERATOR at \[\(3\)
\]
\#\#\# \<== '\(' at line 1 \[f\]
\#\#\# LEX\_NORMAL
\#\#\# Tokener expecting TERM at \[3\)
\]
\#\#\# Saw number in '\)
'
\#\#\# \<== THING\(opval=op\_const\) at line 1 \[3\]
\#\#\# LEX\_NORMAL
\#\#\# Tokener expecting OPERATOR at \[\)
\]
\#\#\# \<== '\)' at line 1 \[3\]
\#\#\# LEX\_NORMAL
\#\#\# Tokener expecting OPERATOR at \[;\]
\#\#\# \<== ';' at line 1 \[3\]
\#\#\# LEX\_NORMAL
\#\#\# Tokener expecting STATE at \[\]
\#\#\# Tokener got EOF
\#\#\# \<== EOF at line 1 \[\]



---
After&#8203;:

\./perl \-DT \-e'$a=$b\+f\(3\)'

\#\#\# 0&#8203;:LEX\_NORMAL/XSTATE "\\n;"
\#\#\# \<== '$'

\#\#\# 1&#8203;:LEX\_NORMAL/XOPERATOR "=$b\+f\(3\)\\n"
\#\#\# Pending identifier '$a'
\#\#\# \<== WORD\(opval=op\_const\) PV\("a"\\0\)

\#\#\# 1&#8203;:LEX\_NORMAL/XOPERATOR "=$b\+f\(3\)\\n"
\#\#\# \<== ASSIGNOP\(ival=op\_null\)

\#\#\# 1&#8203;:LEX\_NORMAL/XTERM "$b\+f\(3\)\\n"
\#\#\# \<== '$'

\#\#\# 1&#8203;:LEX\_NORMAL/XOPERATOR "\+f\(3\)\\n"
\#\#\# Pending identifier '$b'
\#\#\# \<== WORD\(opval=op\_const\) PV\("b"\\0\)

\#\#\# 1&#8203;:LEX\_NORMAL/XOPERATOR "\+f\(3\)\\n"
\#\#\# \<== ADDOP\(ival=op\_add\)

\#\#\# 1&#8203;:LEX\_NORMAL/XTERM "f\(3\)\\n"
\#\#\# \<== '&'

\#\#\# 1&#8203;:LEX\_KNOWNEXT/XOPERATOR "\(3\)\\n"
\#\#\# \<== WORD\(opval=op\_const\) PV\("f"\\0\)

\#\#\# 1&#8203;:LEX\_NORMAL/XOPERATOR "\(3\)\\n"
\#\#\# \<== '\('

\#\#\# 1&#8203;:LEX\_NORMAL/XTERM "3\)\\n"
\#\#\# Saw number in "\)\\n"
\#\#\# \<== THING\(opval=op\_const\) IV\(3\)

\#\#\# 1&#8203;:LEX\_NORMAL/XOPERATOR "\)\\n"
\#\#\# \<== '\)'

\#\#\# 1&#8203;:LEX\_NORMAL/XOPERATOR ";"
\#\#\# \<== ';'

\#\#\# 1&#8203;:LEX\_NORMAL/XSTATE ""
\#\#\# Tokener got EOF
\#\#\# \<== EOF



---
\-\- 
"Do not dabble in paradox\, Edward\, it puts you in danger of fortuitous
wit\." \-\- Lady Croom \- Arcadia

**Comment by p5pRT** [comment]
@iabyn - Status changed from 'open' to 'resolved'
